java/arcs/core/storage/util/SendQueue.kt

```diff
@@ -48,17 +55,55 @@ class SendQueue(
      * Enqueues a send function on the send queue, blocking on its execution until the provided
      * [references] are all available in the [backingStore].
      */
-    suspend fun enqueueBlocking(references: List<Reference>, runnable: suspend () -> Unit) {
-        mutex.withLock {
+    suspend fun enqueueBlocking(
+        references: List<Reference>,
+        runnable: suspend () -> Unit
+    ): Deferred<Unit> {
+        val completableDeferred = CompletableDeferred<Unit>()
+        val queuedBlockName = mutex.withLock {
             val block = "${currentBlock++}"
-            queue.add(PendingSend.Blocking(block, runnable))
-            holdQueue.enqueue(references.map { HoldQueue.Entity(it.id, it.version?.copy()) }) {
+            val holdQueueId = holdQueue.enqueue(
+                references.map { HoldQueue.Entity(it.id, it.version?.copy()) }
+            ) {
                 drain(block)
             }
+            queue.add(
+                PendingSend.Blocking(block, holdQueueId) {
+                    try {
+                        runnable()
+                    } finally {
+                        completableDeferred.complete(Unit)
+                    }
+                }
+            )
+            block
         }
 
         // Additionally, we can drain any non-blocked items.
         if (drainOnEnqueue) drain()
+
+        val localScope = CoroutineScope(Dispatchers.Unconfined + Job())
+        completableDeferred.invokeOnCompletion {
+            if (it is CancellationException) localScope.launch {
```

Comment:
How can this happen?

Reply:
TimeoutCancellationException (extends CancellationException) is what is passed to this completion handler when a `withTimeout` expires.